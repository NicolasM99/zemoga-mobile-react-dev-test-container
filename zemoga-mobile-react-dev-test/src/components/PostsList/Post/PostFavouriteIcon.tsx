import React, { FC } from 'react';

import FontAwesome from '@expo/vector-icons/FontAwesome';
import { GestureResponderEvent, TouchableOpacity, View } from 'react-native';

export interface PostFavouriteIconProps {
  onSetFavourite?: (event: GestureResponderEvent) => void;
  isFavourite?: boolean;
}

const PostFavouriteIcon: FC<PostFavouriteIconProps> = ({
  onSetFavourite,
  isFavourite = false
}: PostFavouriteIconProps) => {
  return (
    <TouchableOpacity onPress={onSetFavourite}>
      <View style={{ flex: 1 }}>
        <FontAwesome name={isFavourite ? 'star' : 'star-o'} size={16} />
      </View>
    </TouchableOpacity>
  );
};

export default PostFavouriteIcon;
