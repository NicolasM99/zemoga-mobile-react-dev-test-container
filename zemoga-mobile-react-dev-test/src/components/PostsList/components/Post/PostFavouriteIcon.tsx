import React, { FC } from 'react';

import FontAwesome from '@expo/vector-icons/FontAwesome';
import { GestureResponderEvent, TouchableOpacity, View } from 'react-native';

export interface PostFavouriteIconProps {
  onSetFavourite?: (event: GestureResponderEvent) => void;
  isFavourite?: boolean;
  isReadOnly?: boolean;
  disabled?: boolean;
}

const PostFavouriteIcon: FC<PostFavouriteIconProps> = ({
  onSetFavourite,
  isFavourite = false,
  isReadOnly = false,
  disabled = false
}: PostFavouriteIconProps) => {
  if (isReadOnly && !isFavourite) return <></>;
  return (
    <View
      style={{
        opacity: disabled ? 0.5 : 1,
        marginLeft: 8,
        position: 'absolute',
        right: 0,
        padding: 16
      }}
    >
      <TouchableOpacity
        testID="postFavouriteIcon"
        disabled={disabled}
        onPress={(e) => (isReadOnly ? null : onSetFavourite?.(e))}
      >
        <FontAwesome color={!isFavourite ? 'rgba(0,0,0, 0.2)' : 'orange'} name="star" size={24} />
      </TouchableOpacity>
    </View>
  );
};

export default PostFavouriteIcon;
