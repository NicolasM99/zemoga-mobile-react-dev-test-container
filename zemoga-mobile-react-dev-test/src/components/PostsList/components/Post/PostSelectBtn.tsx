import React, { FC } from 'react';
import { View } from 'react-native';
import FontAwesome from '@expo/vector-icons/FontAwesome';
import { IPostSelectBtn } from './@types/postSelectBtn';

const PostSelectBtn: FC<IPostSelectBtn> = ({ isSelected = false }: IPostSelectBtn) => {
  return (
    <View style={{ marginRight: 16 }}>
      <FontAwesome
        color={!isSelected ? 'rgba(0,0,0, 0.2)' : 'red'}
        name={!isSelected ? 'circle' : 'check-circle'}
        size={20}
      />
    </View>
  );
};

export default PostSelectBtn;
