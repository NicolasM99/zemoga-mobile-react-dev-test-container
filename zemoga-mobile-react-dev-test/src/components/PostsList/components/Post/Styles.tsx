import { StyleSheet } from 'react-native';

export const postStyles = StyleSheet.create({
  postContainer: {
    flexDirection: 'row',
    width: 'auto',
    paddingLeft: 10,
    margin: 8,
    marginBottom: 4,
    justifyContent: 'space-between',
    alignItems: 'center',
    backgroundColor: 'white',
    paddingVertical: 12,
    paddingRight: 64
  },
  postTitle: {
    fontSize: 16
  }
});
