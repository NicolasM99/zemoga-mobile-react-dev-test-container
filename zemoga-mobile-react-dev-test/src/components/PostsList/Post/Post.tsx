import React, { FC } from 'react';

import { GestureResponderEvent, Text, TouchableOpacity } from 'react-native';

import { postStyles } from './Styles';

type PostProps = {
  onPressPost?: (event: GestureResponderEvent) => void;
};

const Post: FC<PostProps> = ({ onPressPost }) => {
  return (
    <TouchableOpacity onPress={onPressPost} style={postStyles.postContainer}>
      <Text>ICON</Text>
      <Text>qui est esse</Text>
    </TouchableOpacity>
  );
};

export default Post;
