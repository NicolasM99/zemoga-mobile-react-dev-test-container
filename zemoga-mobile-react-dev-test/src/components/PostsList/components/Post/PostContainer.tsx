import React, { FC } from 'react';

import { GestureResponderEvent, TouchableOpacity, View } from 'react-native';

import { postStyles } from './Styles';

export interface PostContainerProps {
  onPressPost?: (event: GestureResponderEvent) => void;
  children: React.ReactNode;
}

const PostContainer: FC<PostContainerProps> = ({ onPressPost, children }: PostContainerProps) => {
  return (
    <TouchableOpacity onPress={onPressPost}>
      <View style={postStyles.postContainer}>{children}</View>
    </TouchableOpacity>
  );
};

export default PostContainer;
