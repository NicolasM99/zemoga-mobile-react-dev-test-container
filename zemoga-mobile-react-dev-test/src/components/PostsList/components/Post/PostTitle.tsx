import React, { FC } from 'react';

import { Text, TouchableWithoutFeedback, View } from 'react-native';

import { postStyles } from './Styles';

export interface PostTitleProps {
  children: React.ReactNode;
}

const PostTitle: FC<PostTitleProps> = ({ children }: PostTitleProps) => (
  <View style={{ flex: 1 }}>
    <Text style={postStyles.postTitle}>{children}</Text>
  </View>
);

export default PostTitle;
