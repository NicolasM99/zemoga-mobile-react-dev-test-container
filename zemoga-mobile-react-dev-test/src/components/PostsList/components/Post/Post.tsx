import React, { FC } from 'react';

import { PostProps } from './@types/post';
import PostContainer from './PostContainer';
import PostFavouriteIcon from './PostFavouriteIcon';
import PostSelectBtn from './PostSelectBtn';
import PostTitle from './PostTitle';

const Post: FC<PostProps> = ({
  onPressPost,
  onSetFavourite,
  isFavourite = false,
  title = '',
  isSelected = false,
  isDeletingItems = false
}) => {
  return (
    <PostContainer onPressPost={onPressPost}>
      {isDeletingItems && <PostSelectBtn isSelected={isSelected} />}
      <PostTitle>{title}</PostTitle>
      {!isDeletingItems && (
        <PostFavouriteIcon onSetFavourite={onSetFavourite} isFavourite={isFavourite} />
      )}
    </PostContainer>
  );
};

export default Post;
