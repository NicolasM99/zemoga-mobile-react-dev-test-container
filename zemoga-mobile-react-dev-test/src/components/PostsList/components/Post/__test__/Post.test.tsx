import React from 'react';

import { fireEvent, render, waitFor } from '@testing-library/react-native';

import Post from '../Post';
describe('<Post /> unit tests', () => {
  const EXPECTED_TITLE = 'post title test';
  it('Should mount the component correctly', () => {
    const screen = render(<Post title={EXPECTED_TITLE} />);
    expect(screen).toMatchSnapshot();
  });

  it('Should match title prop', () => {
    const { getByTestId } = render(<Post title={EXPECTED_TITLE} />);
    waitFor(() => expect(getByTestId('postTitle')).toHaveTextContent(EXPECTED_TITLE));
  });

  it('Should trigger a function on press', () => {
    const customFunction = { pressFunction: (message: string) => message };
    spyOn(customFunction, 'pressFunction');
    const screen = render(
      <Post onPressPost={() => customFunction.pressFunction('pressed')} title={EXPECTED_TITLE} />
    );
    const postItem = screen.getByTestId('postItem');
    fireEvent.press(postItem);
    expect(customFunction.pressFunction).toHaveBeenCalledWith('pressed');
  });

  it('Should trigger a function on pressing favourite icon', () => {
    const customFunction = { pressFunction: (message: string) => message };
    spyOn(customFunction, 'pressFunction');
    const screen = render(
      <Post onSetFavourite={() => customFunction.pressFunction('pressed')} title={EXPECTED_TITLE} />
    );
    const postFavouriteIcon = screen.getByTestId('postFavouriteIcon');
    fireEvent.press(postFavouriteIcon);
    expect(customFunction.pressFunction).toHaveBeenCalledWith('pressed');
  });

  it('Should show select icon and hide favourite icon when isDeletingItems', () => {
    const { getByTestId, queryByTestId } = render(<Post isDeletingItems title={EXPECTED_TITLE} />);
    expect(getByTestId('postSelectIcon')).toBeTruthy();
    expect(queryByTestId('postFavouriteIcon')).not.toBeTruthy();
  });

  it('Should show select icon and show favourite icon when isDeletingItems and isFavourite', () => {
    const { getByTestId } = render(<Post isFavourite isDeletingItems title={EXPECTED_TITLE} />);
    expect(getByTestId('postSelectIcon')).toBeTruthy();
    expect(getByTestId('postFavouriteIcon')).toBeTruthy();
  });

  it("Shouldn't trigger function on press of favourite icon when isDeletingItems", () => {
    const customFunction = { pressFunction: (message: string) => message };
    spyOn(customFunction, 'pressFunction');
    const screen = render(
      <Post
        isFavourite
        onSetFavourite={() => customFunction.pressFunction('pressed')}
        isDeletingItems
        title={EXPECTED_TITLE}
      />
    );
    const postFavouriteIcon = screen.getByTestId('postFavouriteIcon');
    fireEvent.press(postFavouriteIcon);
    expect(customFunction.pressFunction).not.toHaveBeenCalled();
  });

  it('Should show the check-circle icon if isSelected', () => {
    const { getByTestId } = render(<Post isDeletingItems isSelected title={EXPECTED_TITLE} />);
    waitFor(() => expect(getByTestId('postSelectIcon').props.name).toEqual('check-circle'));
  });

  it('Should show the circle icon if isSelected is false', () => {
    const { getByTestId } = render(<Post isDeletingItems title={EXPECTED_TITLE} />);
    waitFor(() => expect(getByTestId('postSelectIcon').props.name).toEqual('circle'));
  });
});
