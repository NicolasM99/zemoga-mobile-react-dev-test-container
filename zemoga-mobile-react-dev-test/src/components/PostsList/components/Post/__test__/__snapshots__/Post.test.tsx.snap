// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Post /> unit tests Should mount the component correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="postItem"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "margin": 8,
        "marginBottom": 4,
        "paddingLeft": 10,
        "paddingRight": 64,
        "paddingVertical": 12,
        "width": "auto",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 16,
          }
        }
        testID="postTitle"
      >
        post title test
      </Text>
    </View>
    <View
      style={
        Object {
          "marginLeft": 8,
          "opacity": 1,
          "padding": 16,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="postFavouriteIcon"
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
